src/Covenant/ASG.hs

```diff
@@ -0,0 +1,232 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+
+-- |
+-- Module: Covenant.ASG
+-- Copyright: (C) MLabs 2025
+-- License: Apache 2.0
+-- Maintainer: koz@mlabs.city, farseen@mlabs.city, sean@mlabs.city
+--
+-- Contains the basic functionality to build up Covenant expressions
+-- programmatically.
+--
+-- @since 1.0.0
+module Covenant.ASG
+  ( -- * Types
+    Id,
+    Arg,
+    Bound,
+    Ref (..),
+    PrimCall (..),
+    ASGNode,
+    ASGBuilder,
+    Scope,
+    ASG,
+
+    -- * Functions
+    emptyScope,
+
+    -- ** Build up expressions
+    lit,
+    prim,
+    arg,
+    bound,
+    app,
+    lam,
+    letBind,
+
+    -- ** Compile an expression
+    toASG,
+  )
+where
+
+import Algebra.Graph.Acyclic.AdjacencyMap
+  ( AdjacencyMap,
+    toAcyclic,
+    vertex,
+  )
+import Algebra.Graph.AdjacencyMap qualified as Cyclic
+import Control.Monad.State.Strict (runState)
+import Covenant.Internal.ASGBuilder
+  ( ASGBuilder (ASGBuilder),
+    ASGBuilderState (ASGBuilderState),
+    app,
+    idOf,
+    lit,
+    prim,
+  )
+import Covenant.Internal.ASGNode
+  ( ASGNode (App, Lam, Let, Lit, Prim),
+    Arg (Arg),
+    Bound (Bound),
+    Id,
+    PrimCall (PrimCallOne, PrimCallSix, PrimCallThree, PrimCallTwo),
+    Ref (ABound, AnArg, AnId),
+  )
+import Data.Bimap (Bimap)
+import Data.Bimap qualified as Bimap
+import Data.Maybe (mapMaybe)
+import Data.Proxy (Proxy (Proxy))
+import GHC.TypeNats (CmpNat, KnownNat, natVal, type (+))
+import Numeric.Natural (Natural)
+
+-- | A Covenant program, represented as an acyclic graph.
+--
+-- @since 1.0.0
+data ASG = ASG (Id, ASGNode) (AdjacencyMap (Id, ASGNode))
```

Comment:
Does this stand for Acyclic Syntax Graph?

Reply:
Yes.